src/devices/bus/isa/areplay.cpp

```diff
@@ -0,0 +1,614 @@
+	// ****************
+	//  WRITE PORT #1
+	// ****************
+	case 0:  // (EEP)ROM bank select ( 1FFF = 8 K chunks).  Range: 0...15 (allows 128 K)
+
+		m_write_port1 = data;
+
+		bank_end = m_current_rom_start + SIZE_ROM_WINDOW - 1;
+		if(data & WRITE_ENABLE_BIT) // detect write to EEPROM area
+		{
+			m_isa->unmap_bank(m_current_rom_start, bank_end);
+			m_isa->install_bank(m_current_rom_start, bank_end,
+				    "banked_flash", &m_banked_flash[ 16 * SIZE_ROM_WINDOW ]
+				          );
+
+			if(data == 0x01 ) // attempt to query manufacturer ID ?
+			{
+				popmessage("EEPROM updates are unemulated!\n[Action Replay]");
+				m_isa->install_bank(m_current_rom_start, bank_end,
+					    "banked_flash", &m_banked_flash[ m_current_rom_page * SIZE_ROM_WINDOW ]
+					 	   );
+			}
+		}
+		else
+		{       // ROM PAGE 0 * IS * USED. MSB BITS (11000011) ARE ALWAYS ZERO:
+			m_current_rom_page = (data & 0x3c) >> 2; // bits 2...5 
+
+			m_isa->unmap_bank(m_current_rom_start, bank_end);
+			m_isa->install_bank(m_current_rom_start, bank_end,
+					    "banked_flash", &m_banked_flash[ m_current_rom_page * SIZE_ROM_WINDOW ]
+				           );
+
+			m_isa->writeprotect_bank(m_current_rom_start, bank_end); // *** write protect ROM bank ****
+		}
+
+		if( ((data & LED_ENABLE_BIT) > 0) && (old_led_state == false) )
+				if( ioport("DEBUG")->read() != 0 ) printf(" [Action Replay LED 0 > 1] ");
+
+		if( ((data & LED_ENABLE_BIT) == 0) && (old_led_state == true) )
+				if( ioport("DEBUG")->read() != 0 ) printf(" [Action Replay LED 1 > 0] ");
+
+		old_led_state = (data & LED_ENABLE_BIT) ? true : false; // keep position of statement 
+
+		break; // (end of) case 0
+
+	// ****************
+	//  WRITE PORT #2
+	// ****************
+	case 1:  // RAM bank select ( FFF = 4K size). Range: 0...31 (128 K possible; amount probed by firmware)
+
+		 if(old_irq_state == false) // Trigger IRQ only if 0 => 1 transition 
+		 {
+                  if( ((data & PORT2_IRQ_ENABLE) > 0 )  )
+		  {
+		     if(m_timer_fired == true)
+		     {	
+			if(m_is_heartbeat_present == false) 
+			   force_irq_to(true);
+		     } 
+		  }
+		 } // old irq state was ZERO.  
+		
+		if( (data & PORT2_IRQ_ENABLE) == 0 ) // driver reactivates stuck IRQ by forcing the bit to 0, then 1
+			force_irq_to(false);
+
+		old_irq_state = (data & PORT2_IRQ_ENABLE) ? true : false; // keep position of statement 
+
+		bank_start = m_current_rom_start + SIZE_ROM_WINDOW;
+		bank_end   = bank_start + SIZE_RAM_BANK - 1;
+
+		m_isa->unmap_bank(bank_start, bank_end);
+
+		m_current_RAM_PAGE = data & 0x1f; // Bits 0..4 
+
+		m_isa->install_bank(bank_start, bank_end , 
+				     "banked_ram", &m_board_ram[ SIZE_RAM_BANK * m_current_RAM_PAGE ]
+			           );
+		break; // (end of) case 1 : RAM bank 
+    }
+}
+
+void isa8_areplay_device::device_timer(emu_timer &timer, device_timer_id tid, int param, void *ptr)
+{
+	switch (tid)
+	{
+		default:
+
+if( ioport("DEBUG")->read() != 0 )
+	printf("\nONE SHOT  **** stopped ****!");
```

Comment:
Don't use `printf` for diagnostic information - use `logerror` or the stuff in `logmacro.h` so it works with the framework.  Also, debugging code needs to follow indentation properly.